frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -60,33 +73,134 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) -> Result<(), MinerError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	match storage.mutate::<_, (), _>(|_| Ok(call.clone())) {
+		Ok(Ok(_)) => Ok(()),
+		Ok(Err(_)) => Err(MinerError::FailedToStoreSolution),
+		Err(_) => unreachable!("mutation function cannot fail; qed"),
+	}
+}
+
+/// Get a saved solution from OCW storage if it exists.
+fn restore_solution<T: Config>() -> Result<Call<T>, MinerError> {
+	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL)
+		.get()
+		.flatten()
+		.ok_or(MinerError::NoStoredSolution)
+}
```

Comment:
```suggestion
fn restore_solution<T: Config>() -> Option<Call<T>> {
	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL)
		.get()
		.flatten()
}
```

This error type was never really used, as you do `or_else` on it later. 

I am also a big fan of keeping error variants minimal, so a plus for that.

Reply:
namely what I mean is that you do `or_else(|_| { ... })`, so the error is never used, so option.

Reply:
The reason I'm using a `Result` instead of an `Option` here is precisely because of the `or_else`. Specifically, that function is defined on an `Option` to take [`F: FnOnce() -> Option<T>`](https://doc.rust-lang.org/std/option/enum.Option.html#method.or_else). That doesn't work for us, because we want to be able to use the `?` operator on `mine_checked_call` and `save_solution`. 

This gives 3 fundamental options:

- What I implemented: `restore_solution` returns a `Result` whose `Err` variant is never exposed
- `restore_solution` returns an `Option`, and we put an `ok_or(MinerError::NeverUsed)` before the `or_else`
- restructure the whole thing into a `match` of some kind

I still think that the first option is the best in this case.

Reply:
I am a fan of the second one with an emphasis that an error variant that is never used doesn't look nice, but your call at this point.